sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala

```diff
@@ -254,24 +255,104 @@ object DecorrelateInnerQuery extends PredicateHelper {
-      union: Union,
+      setOp: LogicalPlan, // Union or SetOperation
       child: LogicalPlan): Seq[Expression] = {
     // The output attributes are always equal to the left child's output
-    assert(union.output.size == child.output.size)
-    val map = AttributeMap(union.output.zip(child.output))
-    conditions.map {
+    assert(setOp.output.size == child.output.size)
+    val map = AttributeMap(setOp.output.zip(child.output))
+    conditions.collect {
       // The left hand side is the domain attribute used in the inner query and the right hand side
       // is the attribute from the outer query. (See comment above in buildDomainAttrMap.)
       // We need to remap the attribute names used in the inner query (left hand side) to account
       // for the different names in each union child. We should not remap the attribute names used
       // in the outer query.
+      //
+      // Note: the reason we can't just use the original joinCond from when the DomainJoin was
+      // constructed is that constructing the DomainJoins happens much earlier than rewriting the
+      // DomainJoins into actual joins, with many optimization steps in
+      // between, which could change the attributes involved (e.g. CollapseProject).
       case EqualNullSafe(left: Attribute, right: Expression) =>
         EqualNullSafe(map.getOrElse(left, left), right)
-      case EqualTo(left: Attribute, right: Expression) =>
-        EqualTo(map.getOrElse(left, left), right)
+    }
+  }
+
+  /**
+   * This is to handle INTERSECT/EXCEPT DISTINCT which are rewritten to left semi/anti join in
+   * ReplaceIntersectWithSemiJoin and ReplaceExceptWithAntiJoin.
+   *
+   * To rewrite the domain join on the right side, we need to remap the attributes in the domain
+   * join cond, using the mapping between left and right sides in the semi/anti join cond.
+   *
+   * After DecorrelateInnerQuery, the domain join conds reference the output names of the
+   * INTERSECT/EXCEPT, which come from the left side. When rewriting the DomainJoin in the
+   * right child, we need to remap the domain attribute names to account for the different
+   * names in the left vs right child, similar to pushDomainConditionsThroughSetOperation.
+   * But after the rewrite to semi/anti join is performed, we instead need to do the remapping
+   * based on the semi/anti join cond which contains equi-joins between the left and right
+   * outputs.
+   *
+   * The domain conds will be like leftInner <=> outer, and the semi/anti join cond will be like
+   * leftInner <=> rightInner. We add additional domain conds rightInner <=> outer which are used
+   * to rewrite the right-side DomainJoin.
+   *
+   * Example: Take a query like:
+   * select * from t0 join lateral (
+   *   select a from t1 where b < x
+   *   intersect distinct
+   *   select b from t2 where c < y)
+   *
+   * Step 1:
+   * We have outer table with attributes [x#1, y#2] and after introducing DomainJoins the subquery
```

Comment:
nit: may be point out explictly from what table (like t0, t1 ,t2) the attributes are coming from. Just to make the comment a bit easier to read.